### PR TITLE
Introduce Backend resource and remove BackendPolicy support

### PR DESCRIPTION
**Pull request by Amila-Rukshan**
## Purpose
$subject

## Task
https://github.com/wso2/apk/issues/799

### CODE REVIEW

File: helm-charts/templates/control-plane/admin-ds/admin-domain-api-httproute.yaml [hunk 1]
```diff
@@ -38,12 +38,7 @@ spec:
           type: "ReplaceFullPath"
           replaceFullPath: "/api/am/admin/policies/search"
     backendRefs:
-    - weight: 1
-      group: ""
-      kind: "Service"
-      name: {{ template "apk-helm.resource.prefix" . }}-admin-ds-service
-      namespace: {{ .Release.Namespace }}
-      port: 9443
+    -
```
Comment (Amila-Rukshan) [inline helm-charts/templates/control-plane/admin-ds/admin-domain-api-httproute.yaml:41]: missing backend

**Review by tharindu1st — DISMISSED** [review]

**Review by tharindu1st — APPROVED** [review]

File: adapter/pkg/logging/logging_constant.go [hunk 2]
```diff
@@ -174,11 +183,6 @@ var Mapper = map[int]ErrorDetails{
 		Message:   "Error retrieving ref CRs for API in namespace : %s, %v",
 		Severity:  TRIVIAL,
 	},
-	error2621: {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/logging/logging_constant.go:177]: can you remove this error codes from the above constant list too?

File: adapter/pkg/logging/logging_constant.go [hunk 3]
```diff
@@ -265,6 +264,51 @@ var Mapper = map[int]ErrorDetails{
 		Message:   "Unable to create webhook for APIPolicy: %v",
 		Severity:  BLOCKER,
 	},
+	error2639: {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/logging/logging_constant.go:267]: may be you can use above removed errorcodes for new entries.

File: adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go:2]:
```suggestion
 *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go [hunk 5]
```diff
@@ -17,16 +17,139 @@
...
+// SecurityConfig defines enpoint security configurations
+type SecurityConfig struct {
+	Type  string              `json:"type,omitempty"`
+	Basic BasicSecurityConfig `json:"basic,omitempty"`
+}
+
+// BasicSecurityConfig defines basic security configurations
+type BasicSecurityConfig struct {
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go:95]: need to change this to a secret.

File: adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go [hunk 6]
```diff
@@ -17,16 +17,139 @@
...
+}
+
+// BackendStatus defines the observed state of Backend
+type BackendStatus struct {
+	// Status denotes the state of the Backend in its lifecycle.
+	// Possible values could be Accepted, Invalid, Deploy etc.
+	//
+	//
+	// +kubebuilder:validation:MinLength=4
+	Status string `json:"status"`
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go:105]: do we update the status in controller? else remove additional fields, we can add them once we support them.

File: adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go (hunk 5 above)
Comment (Amila-Rukshan) [reply adapter/pkg/operator/apis/dp/v1alpha1/backend_types.go]: updated it with CR configuration as this:
```
  security:
  - type: Basic
    basic:
      secretRef: 
        name: credentials-secret
        usernameKey: usr
        passwordKey: pwd
```

**Review by tharindu1st — APPROVED** [review]

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**